next/broker/celery_app/tasks.py

```diff
@@ -7,20 +7,63 @@
 import json
 import traceback
 import numpy
-from next.constants import DEBUG_ON
 import hashlib
 
-# import next.logging_client.LoggerHTTP as ell
-from next.database_client.DatabaseAPI import DatabaseAPI
-db = DatabaseAPI()
-from next.logging_client.LoggerAPI import LoggerAPI
-ell = LoggerAPI()
 import next.utils
 import next.constants
-import next.apps.Butler as Butler
+from next.constants import DEBUG_ON
+
+from next.database_client.DatabaseAPI import DatabaseAPI, DatabaseException
+from next.logging_client.LoggerAPI import LoggerAPI
+
+from next.apps.Butler import Butler
 import next.lib.pijemont.verifier as verifier
 
-Butler = Butler.Butler
+db, ell = None, None
+
+def worker_connect_db():
+    next.utils.debug_print("Initializing worker database connection")
+    backoff_dt = 0.01
+    while True:
+        try:
+            db, ell = DatabaseAPI(), LoggerAPI()
+            break
+        except DatabaseException as e:
+            next.utils.debug_print("Failed to connect to database ({}), retrying in {}s".format(e, backoff_dt))
+            time.sleep(backoff_dt)
+            backoff_dt *= 2
+
+            # If we've tried too many times, make the database failure /loud/.
+            if backoff_dt > 1:
+                raise e
+
+    return db, ell
+
+# if we're not using celery, just initialize the database globally
+if next.constants.CELERY_ON:
+    db, ell = worker_connect_db()
+
+# runs for each worker process spawned by Celery
+# we initialize the DatabaseAPI per worker here
+@celery.signals.worker_process_init.connect
+def on_connect(**kwargs):
+    global db, ell
+    # make sure every worker has a different random seed (very important for randomized algorithms)
+    numpy.random.seed()
```

Comment:
(my experience with setting random seeds on multiple processes leads me to doubt this seed is different for each worker)